### PR TITLE
FiLM geometry-conditioned vol output head to close OOD vol_p gap

### PR DESCRIPTION
**Pull request by morganmcg1**
## Hypothesis

The structural val→test `volume_pressure_rel_l2_pct` gap (~+7.6pp) is confirmed as a data distribution problem: the 50 test geometries lie in a different region of geometry space than the 400 train/val geometries. Every model-side intervention so far has failed to close this gap.

**Hypothesis**: The model's volume pressure predictions have poor generalization because the output head applies a single global learned mapping regardless of geometry type. If we condition the volume output MLP on a **global geometry embedding** (derived from surface/body geometry features), the model can learn geometry-specific output scalings that transfer better to out-of-distribution test geometries.

**Mechanism**: Extract a global geometry descriptor `g` from the surface point features (e.g., mean-pooled surface hidden states, shape `[B, D]`). Use `g` to modulate the volume output head via FiLM (Feature-wise Linear Modulation): `vol_out = scale(g) * vol_features + bias(g)` where `scale` and `bias` are small MLPs. This allows the network to apply geometry-aware output scaling rather than a fixed projection, which may generalize better to OOD geometries.

**Reference**: Perez et al. "FiLM: Visual Reasoning with a General Conditioning Layer" (AAAI 2018). FiLM conditioning has been widely used in NeRF/implicit function work to condition outputs on global context.

### 1. Add CLI flag

In `target/train.py`:
```python
parser.add_argument("--use-film-vol-conditioning", action="store_true", default=False,
                    help="Condition volume output head on global surface geometry embedding via FiLM")
```

### 2. Implement FiLM conditioning module

In the model file, add a `FiLMConditioner` module:

```python
class FiLMConditioner(nn.Module):
    """
    Generates FiLM scale+bias from a global geometry embedding.
    g: [B, D_global]  ->  scale: [B, 1, D_hidden], bias: [B, 1, D_hidden]
    """
    def __init__(self, d_global: int, d_hidden: int):
        super().__init__()
        self.scale_net = nn.Sequential(
            nn.Linear(d_global, d_hidden),
            nn.SiLU(),
            nn.Linear(d_hidden, d_hidden),
        )
        self.bias_net = nn.Sequential(
            nn.Linear(d_global, d_hidden),
            nn.SiLU(),
            nn.Linear(d_hidden, d_hidden),
        )
        # Initialize to identity (scale=1, bias=0)
        nn.init.zeros_(self.scale_net[-1].weight)
        nn.init.ones_(self.scale_net[-1].bias)
        nn.init.zeros_(self.bias_net[-1].weight)
        nn.init.zeros_(self.bias_net[-1].bias)

    def forward(self, vol_features, geometry_embedding):
        # vol_features: [B, N_vol, D_hidden]
        # geometry_embedding: [B, D_global]
        scale = self.scale_net(geometry_embedding).unsqueeze(1)  # [B, 1, D_hidden]
        bias  = self.bias_net(geometry_embedding).unsqueeze(1)   # [B, 1, D_hidden]
        return vol_features * scale + bias
```

### 3. Compute global geometry embedding

In the forward pass, after processing surface tokens, compute a global geometry embedding via mean pooling:
```python
# surf_hidden: [B, N_surf, D]  — surface hidden states after transformer
geometry_embedding = surf_hidden.mean(dim=1)  # [B, D]
```

If surface hidden states are not separately accessible, use the global CLS/readout token if one exists, or pool the volume encoder hidden states: `geometry_embedding = vol_hidden.mean(dim=1)`.

### 4. Apply FiLM before the volume output projection

Locate the volume pressure output head (the linear projection that maps volume features to scalar pressure predictions). Insert FiLM conditioning immediately before it:

```python
# Before output projection in volume branch:
if self.use_film_vol_conditioning:
    vol_hidden = self.film_conditioner(vol_hidden, geometry_embedding)
# Then: vol_pred = self.vol_output_head(vol_hidden)
```

Add to model `__init__`:
```python
if use_film_vol_conditioning:
    self.film_conditioner = FiLMConditioner(d_global=hidden_dim, d_hidden=hidden_dim)
```

### 5. Run command

```bash
cd target/ && torchrun --nproc_per_node=8 train.py \
  --epochs 35 \
  --lr 1e-4 \
  --weight-decay 0.005 \
  --optimizer lion \
  --grad-clip-norm 1.0 \
  --train-volume-points 16384 \
  --eval-volume-points 65536 \
  --use-film-vol-conditioning \
  --wandb-group askeladd-film-vol-conditioning \
  --kill-thresholds '{"ep3_val_abupt": 8.0, "ep3_val_vol_p": 5.0, "ep5_val_abupt": 7.5, "ep10_val_abupt": 7.2, "ep15_val_abupt": 6.80, "ep20_val_abupt": 6.70, "ep25_val_abupt": 6.65, "ep30_val_abupt": 6.60}'
```

### 6. Kill gates

- EP3: val_abupt ≤ 8.0%, val_vol_p ≤ 5.0%
- EP5: val_abupt ≤ 7.5%
- EP10: val_abupt ≤ 7.2%
- EP15: val_abupt ≤ 6.80%
- EP20: val_abupt ≤ 6.70%
- EP25: val_abupt ≤ 6.65%
- EP30: val_abupt ≤ 6.60%

## Baseline

Current best metrics (Wave SOTA — PR #740, W&B run `5x8wofzm`):
| Metric | val | test |
|--------|-----|------|
| abupt | ~6.46% | 7.5195% |
| vol_p | ~6.29% | **10.758%** |
| surf_p | ~3.30% | 3.881% |
| wall | ~4.84% | 7.061% |

Public AB-UPT targets: vol_p=6.08%, surf_p=3.82%, wall=7.29%

Structural val→test vol_p gap: **+7.6pp** — this is the primary target.

W&B baseline: https://wandb.ai/wandb-applied-ai-team/senpai-v1-drivaerml-ddp8/runs/5x8wofzm

Reproduce baseline:
```bash
cd target/ && torchrun --nproc_per_node=8 train.py \
  --epochs 35 --lr 1e-4 --weight-decay 0.005 --optimizer lion \
  --grad-clip-norm 1.0 --train-volume-points 16384 --eval-volume-points 65536
```

## Expected outcome

If the FiLM conditioning improves OOD generalization: test_vol_p should drop below 10.758%. The geometry-conditioned output head should be better able to adapt its prediction scale/offset to geometry types not seen during training. If this works, it suggests more aggressive geometry conditioning (e.g., conditioning multiple transformer layers) would be even more effective.

### CODE REVIEW

**Comment by morganmcg1** [comment]
ADVISOR: No acknowledgment on PR #1045 (FiLM geometry-conditioned vol output head) yet.

Please acknowledge this assignment and:
1. Begin implementing the FiLMConditioner module as specified in the PR body
2. Post W&B run ID once the run is launched

**EP3 kill gate**: val_abupt ≤8.0% AND val_vol_p ≤5.0%.

Key implementation: FiLMConditioner generates scale+bias from mean-pooled surface hidden states, applied before the vol output projection. Initialize to identity (scale=1, bias=0) for stable convergence from the SOTA baseline. Post EP1 and EP2 val metrics inline.

**Comment by morganmcg1** [comment]
ADVISOR: SECOND ESCALATION — No response received to my previous comment. This PR has been sitting with zero student activity since it was assigned. The FiLM conditioning experiment (PR #1045) needs to start immediately.

To recap the assignment:
- Implement `FiLMConditioner` that computes a global surface geometry embedding (mean-pooled surface node features) and applies feature-wise linear modulation to the vol output head: `scale(g) * vol_features + bias(g)`
- `scale_net` and `bias_net` are two-layer MLPs initialized so scale=1, bias=0 at the start
- Flag: `--use-film-vol-conditioning`
- W&B group: `askeladd-film-vol-conditioning`
- Kill gate EP3: val_abupt ≤ 8.0% AND val_vol_p ≤ 5.0%

Acknowledge this comment and post the W&B run ID once launched. This is urgent.

**Comment by morganmcg1** [comment]
ADVISOR: THIRD ESCALATION — PR #1045 (FiLM geometry-conditioned vol output head, askeladd) still has zero student activity.

Post an acknowledgment and launch the run immediately. The FiLMConditioner implementation is described in full detail in the PR body: ~40 lines of code. If you are blocked, say so now.

Continued non-response will result in this PR being closed and re-assigned.

**Comment by morganmcg1** [comment]
ADVISOR: Closing PR #1045 because ADVISOR: FINAL — Closing PR #1045 (FiLM geometry-conditioned vol decoder, askeladd). Four escalation comments have been posted since assignment with zero student response. This experiment is being closed and will be reassigned to another student. GPU time cannot be left idle..
